internal/tofu/node_resource_abstract_instance.go

```diff
@@ -105,6 +108,84 @@ func (n *NodeAbstractResourceInstance) References() []*addrs.Reference {
 	return nil
 }
 
+func (n *NodeAbstractResourceInstance) ResolveProvider(ctx EvalContext) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	log.Printf("[TRACE] Resolving provider key for %s", n.Addr)
+
+	if n.Config != nil && n.Config.ProviderConfigRef != nil && n.Config.ProviderConfigRef.KeyExpression != nil {
+		if n.ResolvedProviderKeyExpr != nil {
+			// Module key and resource key are required. This is not allowed!
+			diags = diags.Append(fmt.Errorf("TODO better message: provider key required in both module and resource"))
+			return diags
+		}
+
+		keyExpr := n.Config.ProviderConfigRef.KeyExpression
+		keyData := ctx.InstanceExpander().GetResourceInstanceRepetitionData(n.Addr)
+		keyScope := ctx.EvaluationScope(nil, nil, keyData)
+
+		keyVal, keyDiags := keyScope.EvalExpr(keyExpr, cty.DynamicPseudoType)
+		diags = diags.Append(keyDiags)
+		if keyDiags.HasErrors() {
+			return diags
+		}
+
+		if keyVal.Type() == cty.String {
+			n.ResolvedProviderKey = addrs.StringKey(keyVal.AsString())
```

Comment:
Above comments about unknown, null, and marked values seem to apply to this one too.

Reply:
[846c114](https://github.com/opentofu/opentofu/pull/2105/commits/846c114ae67fb59ca0ed2dd1629a716da6c7dae5)